Profile-ModelarDB-Compression/main.py

```diff
@@ -0,0 +1,118 @@
+""" A script that computes which part of modelardbd's compressed storage format
+    uses the most storage at both the logical database, table, and column level.
+"""
+
+import glob
+import sqlite3
+import sys
+import tempfile
+from os import path
+from collections import defaultdict
+
+from pyarrow import parquet
+
+
+class ColumnMetadata:
+    """Represents the metadata of a logical column."""
+
+    def __init__(self, name, index, error_bound):
+        self.name = name.upper()
+        self.index = index
+        self.error_bound = error_bound
+
+    def __str__(self):
+        return f"{self.name}(error bound {self.error_bound}%)"
+
+
+def read_metadata_for_table(data_folder, table_name):
+    """Read the metadata for a table from the SQLite metadata database."""
+    metadata_database = path.join(data_folder, "metadata.sqlite3")
+    connection = sqlite3.connect(metadata_database)
+    cursor = connection.cursor()
+
+    cursor.execute(
+        (
+            "SELECT column_name, column_index, error_bound "
+            "FROM model_table_field_columns "
+            f"WHERE table_name = '{table_name}'"
+        )
+    )
+
+    table_metadata = {}
+    for row in cursor:
+        column_metadata = ColumnMetadata(row[0], row[1], row[2])
+        table_metadata[column_metadata.index] = column_metadata
+
+    cursor.close()
+    connection.close()
+    return table_metadata
+
+
+def compute_size_of_table_in_bytes_when_stored_as_parquet(table):
+    """Compute the size of table when stored as an Apache Parquet file."""
+    temporary_file = tempfile.NamedTemporaryFile()
+
+    # Zstandard is used for compression to more closely match modelardbd.
+    parquet.write_table(table, temporary_file, compression="zstd")
+    temporary_file.flush()
+    return path.getsize(temporary_file.name)
+
+
+def print_size_in_mib(name, size_in_bytes):
+    """Format and print the size of the entity with name in mebibytes."""
+    size_in_mib = size_in_bytes / 1024 / 1024
+    print(f"  {name:15} {size_in_mib:10.5f} MiB")
```

Comment:
Is it intentional that there are 2 leading spades?

Reply:
Yes it it for formatting of the output.